### DISCUSSION TITLE
Add treesitter integration to `mini.pairs`

### DISCUSSION THREAD

**Discussion by SGauvin** [nvim-mini/mini.nvim#740] [Ideas]
It would be awesome if we could easily integrate treesitter with `mini.pairs`!
I think treesitter is ubiquitous enough where it would make sense, and could add some very useful functionality.

A usecase that I think is very common is people wanting to enable/disable certain pairs based on the current treesitter node.
For example, I think most people would want to disable the single quote pairing when writing a comment, or even perhaps when they are in a string.

I think providing a simple API that lets a user disable a pairing in specific nodes would be very very useful :)

nvim-autopairs has something similar: https://github.com/windwp/nvim-autopairs?tab=readme-ov-file#treesitter

**Comment by echasnovski** [comment]
Thanks for the suggestion!

Although a neat idea, I don't think this will happen in 'mini.pairs'. Its main purpose is to be as simple and fast as possible. Having to check for tree-sitter context each time a pair character is typed (which is quite frequent), goes against this design decision.

Right now there is an only mechanism to decide whether a typed character needs special treatment, and that is [`neigh_pattern`](https://github.com/echasnovski/mini.nvim/blob/c3b37f4afb933141a750b60b86fb6ef6b50d0e61/doc/mini-pairs.txt#L137-L139). For single quote it already is by default [such that only single one is inserted after a Latin letter](https://github.com/echasnovski/mini.nvim/blob/c3b37f4afb933141a750b60b86fb6ef6b50d0e61/doc/mini-pairs.txt#L121). This looks like a more reasonable compromise for this problem.

**Reply by SGauvin** [reply]
@echasnovski I understand!

I feel like there are a lot of situation where a pair char shouldn't be doubled.
For example in Rust, when using lifetimes, the single quote is used a lot!
Eg:
```rust
impl<'a> WatchablePath<'a> {
    pub fn recursive(path: &'a Path) -> Self {
        Self {
            path,
            mode: RecursiveMode::Recursive,
        }
    }
}
```

I feel like tree-sitter would be a more robust solution than modifying the neigh_pattern for single quotes?

**Reply by echasnovski** [reply]
> For example in Rust, when using lifetimes, the single quote is used a lot!

Then my suggestion would be to disable single quote pair for `rust` filetype.

> I feel like tree-sitter would be a more robust solution than modifying the neigh_pattern for single quotes?

Probably yes, but at the cost of significantly more computations done quite frequently. Which feels like a too big price to pay for this.

**Reply by SGauvin** [reply]
Oh well that's a bummer,
Thank you for your time :)
